File: caesar/caesar.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <cs50.h>

/*
Caesar.c
Program that encrypts messages using Caesar's cipher.
This program accepts a command-line argument of a non-negative integer and raises an error if it is not included
*/
int main(int argc,  char* argv[])
{
    //argc = argument count
    //argv = argument vector
    //atoi = converts numeric strings to digits

    if (argc != 2)
    {
        printf("Usage: ./caesar key");
        return 1;
    }
    else if (!isdigit(*argv[1]))
    {
        printf("Usage: ./caesar key");
        return 1;
    }

    int key = atoi(argv[1]);
    if (key < 0)
    {
        printf("Non-negative integers only");
        return 1;
    }


    //Get user input
    string text = get_string("plaintext: ");

    //creating arrays(strings)
    string plainText;
    string cipharText;

    plainText = text;
    cipharText = plainText;


    for(int i = 0, n= strlen(plainText); i < n; i++)
    {
        if (isalpha(plainText[i]))
        {
            if (isupper(plainText[i]))
            {
                int pi = plainText[i] - 'A'; //65 => ascii value for "A"
                int ci = (pi + key) % 26;

                cipharText[i] = ci + 'A';
            }
            else if (islower(plainText[i]))
            {
                int pi = plainText[i] - 'a'; //97 =>ascii value for "a"
                int ci = (pi + key) % 26;

                cipharText[i] = ci + 'a';
            }
        }
        else
        {
            cipharText[i] = plainText[i];
        }
    }
    printf("ciphertext: %s\n", cipharText);
    return 0;
}

File: caesar/caesar.py
"""
This is an encryption-decryption  process where a sentence or word is given and they are encrypted based  
on the number of steps (k) required.
In doing this python has a string module that containsthe ASCII encodings of letters 
(those from the english alphabets inclusive).
Here  we are required to make the file accept a command line argument.
"""

import string, sys

def main(decrypt=False, encrypt=False):
    try:
        key = int(sys.argv[1])
        if key < 1:
            raise ValueError ("Negative key not provided")
    except (ValueError, IndexError):
        print("Usage: ./caesar k")
        sys.exit(1)

    if encrypt:
        plainText = input("Plaintext: ")
        cipharText = caesar_encrypt(plainText, key)
        print(f"{plainText}: {cipharText}")
    if decrypt:
        cipharText = input("Ciphartext: ")
        plainText = caesar_decrypt(cipharText, key)   
        print(f"{cipharText} : {plainText}")

def caesar_encrypt(text, key):
    text= list(text)
    n_alphabets = len(string.ascii_lowercase) #or string.ascii_uppercase

    #in ascii integer of 'A' is 65 while that  of 'a' is 97
    for i, character in enumerate(text):
        if character.isalpha():
            if character.isupper():
                #pi = ord(character) - ord("A") if upper or ord("a") if lower
                #ci = pi + key ci => ciphartext
                text[i] = chr((ord(character) - ord("A") + key) % n_alphabets + ord("A"))
            else:
                text[i] = chr((ord(character) - ord("a") + key) % n_alphabets + ord("a"))
        else:
            text[i] = character
    text = ''.join(text)
    return text

def caesar_decrypt(text, key):
    text = list(text)
    n_alphabets = len(string.ascii_lowercase) #or string.ascii_uppercase

    #in ascii integer of 'A' is 65 while that  of 'a' is 97
    for i, character in enumerate(text):
        if character.isalpha():
            if character.isupper():
                #pi = ord(character) - ord("A") if upper or ord("a") if lower
                #ci = pi + key ci => ciphartext
                text[i] = chr((ord(character) - ord("A") - key) % n_alphabets + ord("A"))
            else:
                text[i] = chr((ord(character) - ord("a") - key) % n_alphabets + ord("a"))
        else:
            text[i] = character
    text = ''.join(text)
    return text


if __name__ == "__main__":
    main(decrypt=True)